File: endpoint/schedule/internal/generate_rounds.go
package internal

import (
	"fmt"
	"log/slog"

	"github.com/yinloo-ola/tournament-manager/model"
)

func GenerateRoundsForTournament(tournament model.Tournament) (model.Tournament, error) {
	for i, category := range tournament.Categories {
		for g, grp := range category.Groups {
			rounds := generateGroupRounds(grp.EntriesIdx, category.DurationMinutes)
			if len(category.Groups[g].Rounds) == 0 {
				category.Groups[g].Rounds = rounds
			} else {
				if len(category.Groups[g].Rounds) != len(rounds) {
					return tournament, fmt.Errorf("number of rounds for group %d is not equal", g+1)
				}
				for j := range category.Groups[g].Rounds {
					for k := range category.Groups[g].Rounds[j] {
						category.Groups[g].Rounds[j][k].Entry1Idx = rounds[j][k].Entry1Idx
						category.Groups[g].Rounds[j][k].Entry2Idx = rounds[j][k].Entry2Idx
					}
				}
			}
		}

		// Always generate knockout rounds based on the current NumQualifiedPerGroup
		koRounds, err := generateKnockoutRounds(category.Groups, category.NumQualifiedPerGroup)
		if err != nil {
			return tournament, fmt.Errorf("generate knock out rounds for category %s failed: %w", category.ShortName, err)
		}
		if len(koRounds) != len(category.KnockoutRounds) {
			category.KnockoutRounds = koRounds
		} else {
			for j := range category.KnockoutRounds {
				if len(category.KnockoutRounds[j].Matches) != len(koRounds[j].Matches) {
					category.KnockoutRounds[j].Matches = koRounds[j].Matches
				} else {
					for k := range category.KnockoutRounds[j].Matches {
						category.KnockoutRounds[j].Matches[k].Entry1Idx = koRounds[j].Matches[k].Entry1Idx
						category.KnockoutRounds[j].Matches[k].Entry2Idx = koRounds[j].Matches[k].Entry2Idx
					}
				}
			}
		}
		tournament.Categories[i] = category
	}
	return tournament, nil
}

func swapRoundWithPlayersToEnd(rounds [][]model.Match, player1, player2 int) {
	roundIdx := -1
	for i, round := range rounds {
		if i == len(rounds)-1 {
			continue
		}
		if roundContains(round, player1, player2) {
			roundIdx = i
		}
	}
	if roundIdx >= 0 {
		rounds[roundIdx], rounds[len(rounds)-1] = rounds[len(rounds)-1], rounds[roundIdx]
	}
}

func roundContains(round []model.Match, player1Idx, player2Idx int) bool {
	for _, match := range round {
		if match.Entry1Idx == player1Idx && match.Entry2Idx == player2Idx {
			return true
		} else if match.Entry1Idx == player2Idx && match.Entry2Idx == player1Idx {
			return true
		}
	}
	return false
}

func isRoundValid(rounds [][]model.Match, numMatches int, numMatchesPerRound int) bool {
	totalMatchCount := 0
	for _, matches := range rounds {
		if len(matches) != numMatchesPerRound {
			slog.Error("num of matches wrong!", "expected", numMatchesPerRound, "gotten", len(matches))
			return false
		}
		totalMatchCount += len(matches)
	}

	return totalMatchCount == numMatches
}

func nextPowerOfTwo(x int) int {
	if x <= 1 {
		return 1
	}
	x--
	x |= x >> 1
	x |= x >> 2
	x |= x >> 4
	x |= x >> 8
	x |= x >> 16
	x |= x >> 32
	x++
	return x
}

func generateKnockoutRounds(groups []model.Group, numQualifiedPerGroup int) ([]model.KnockoutRound, error) {
	for _, group := range groups {
		if len(group.EntriesIdx) < numQualifiedPerGroup {
			return nil, fmt.Errorf("not enough players")
		}
	}

	qualifiedPlayersNum := len(groups) * numQualifiedPerGroup
	firstRound := nextPowerOfTwo(qualifiedPlayersNum)
	numByes := firstRound - qualifiedPlayersNum
	numMatches := (firstRound / 2) - numByes

	koRounds := make([]model.KnockoutRound, 0)

	round := firstRound
	for ; round >= 2; round = round / 2 {
		if round == firstRound {
			matches := make([]model.Match, numMatches)
			for i := range matches {
				matches[i] = model.Match{
					Entry1Idx: model.EntryEmptyIdx,
					Entry2Idx: model.EntryEmptyIdx,
				}
			}
			koRound := model.KnockoutRound{
				Round:   round,
				Matches: matches,
			}
			koRounds = append(koRounds, koRound)
			continue
		}

		matches := make([]model.Match, round/2)
		for i := range matches {
			matches[i] = model.Match{
				Entry1Idx: model.EntryEmptyIdx,
				Entry2Idx: model.EntryEmptyIdx,
			}
		}
		koRound := model.KnockoutRound{
			Round:   round,
			Matches: matches,
		}
		koRounds = append(koRounds, koRound)
	}

	return koRounds, nil
}

func generateGroupRounds(entriesIdx []int, matchDurationMinutes int) [][]model.Match {
	if len(entriesIdx) < 2 {
		return nil
	}

	numPlayers := len(entriesIdx)
	numMatches := (numPlayers * (numPlayers - 1)) / 2
	numMatchesPerRound := numPlayers / 2
	numRounds := numMatches / numMatchesPerRound

	if numPlayers%2 == 1 {
		entriesIdx = append(entriesIdx, model.EntryByeIdx)
		numPlayers++
	}

	rounds := make([][]model.Match, numRounds)
	indices := make([]int, numPlayers)
	for r := 0; r < numRounds; r++ {
		rounds[r] = getRoundMatches(r, entriesIdx, matchDurationMinutes, indices)
	}

	if !isRoundValid(rounds, numMatches, numMatchesPerRound) {
		slog.Error("generateGroupRounds encounter error", "rounds", rounds, "numMatches", numMatches)
		panic("generateGroupRounds encounter error")
	}

	if len(entriesIdx) > 2 {
		swapRoundWithPlayersToEnd(rounds, entriesIdx[1], entriesIdx[2])
	}
	return rounds
}

func getRoundMatches(round int, entriesIdx []int, matchDurationMinutes int, indices []int) []model.Match {
	getRoundPlayersIndices(round, len(entriesIdx), indices)
	matches := make([]model.Match, 0, len(entriesIdx)/2)

	for i := 0; i < len(indices); i += 2 {
		ind1, ind2 := indices[i], indices[i+1]
		if ind2 < ind1 {
			ind1, ind2 = ind2, ind1
		}
		p1 := entriesIdx[ind1]
		p2 := entriesIdx[ind2]
		if p1 == model.EntryByeIdx || p2 == model.EntryByeIdx {
			continue
		}
		match := model.Match{
			Entry1Idx:       p1,
			Entry2Idx:       p2,
			DurationMinutes: matchDurationMinutes,
		}
		matches = append(matches, match)
	}
	return matches
}

func getRoundPlayersIndicesWithRotation(round, numPlayers int, sliceForRotation, res []int) {
	if numPlayers%2 == 1 {
		panic("num of players should be even")
	}
	if round+1 >= numPlayers {
		panic("invalid number of rounds or numPlayers")
	}

	res[0] = 0

	indices := make([]int, len(sliceForRotation))
	copy(indices, sliceForRotation)
	ind := 0

	rotateInPlace(indices, round)

	// for all second players starting from the first
	for i := 1; i < numPlayers; i += 2 {
		res[i] = indices[ind]
		ind++
	}

	// for all first
	for i := numPlayers - 2; i > 0; i -= 2 {
		res[i] = indices[ind]
		ind++
	}
}

// getRoundPlayersIndices computes the indices of players for a given round in a round-robin tournament.
// This function fills the `res` slice with indices of players for the specified `round`.
// It expects an even number of players and the number of rounds should not exceed the number of players.
// The `res` slice will contain the calculated indices, with `res[0]` always being 0.
// The function panics if the input constraints are not met.
// Example:
// For 4 players (0, 1, 2, 3):
//   - Round 0: res = [0, 1, 2, 3]
//   - Round 1: res = [0, 3, 1, 2]
//   - Round 2: res = [0, 2, 3, 1]
func getRoundPlayersIndices(round, numPlayers int, res []int) {
	if numPlayers%2 == 1 {
		panic("num of players should be even")
	}
	if round+1 >= numPlayers {
		panic("invalid number of rounds or numPlayers")
	}

	res[0] = 0

	// Iterate through all players except player 0 (who stays fixed at position 0)
	for i := 1; i < numPlayers; i++ {
		var newPos int

		if i%2 == 0 { // For players at even positions (2, 4, 6, etc.)
			// Move forward by 2*round positions for each round
			// This creates a clockwise rotation pattern for even-positioned players
			newPos = i + 2*round

			// If the new position exceeds the valid range (0 to numPlayers-1)
			if newPos >= numPlayers {
				// Apply "bouncing" logic when we hit the boundary
				// This formula reflects the position back from the boundary
				// Example: With 8 players, if newPos = 9, it becomes 2*8-9-1 = 6
				newPos = 2*numPlayers - newPos - 1

				// Handle double reflection case (when the reflection itself is out of bounds)
				// This can happen with larger round numbers
				if newPos < 0 {
					// Second reflection: if we bounce back past 0, reflect again
					// The formula -(newPos - 1) ensures we stay within valid range
					newPos = -(newPos - 1)
				}
			}
		} else { // For players at odd positions (1, 3, 5, etc.)
			// Move backward by 2*round positions for each round
			// This creates a counterclockwise rotation pattern for odd-positioned players
			newPos = i - 2*round

			// If the new position is negative (below the valid range)
			if newPos < 0 {
				// Apply "bouncing" logic when we hit the lower boundary
				// This formula reflects the position back from the boundary
				// Example: With 8 players, if newPos = -1, it becomes -(-1-1) = 2
				newPos = -(newPos - 1)

				// Handle double reflection case (when the reflection exceeds upper bound)
				if newPos >= numPlayers {
					// Second reflection: if we bounce past numPlayers, reflect again
					// The formula 2*numPlayers - newPos - 1 ensures we stay within valid range
					newPos = 2*numPlayers - newPos - 1
				}
			}
		}

		res[i] = newPos
	}
}

func generateSlice(n int) []int {
	if n%2 == 0 {
		panic("only odd slice supported")
	}
	// Create a slice to hold the result
	result := make([]int, n)

	// Calculate the midpoint (N/2 + 1) for odd numbers
	midpoint := n/2 + 1

	// Fill the first half of the slice with odd numbers starting from 1
	for i := 0; i < midpoint; i++ {
		result[i] = 2*i + 1
	}

	// Fill the second half with even numbers in reverse order starting from 2
	for i := midpoint; i < n; i++ {
		result[i] = 2 * (n - i)
	}

	return result
}

func rotateInPlace(s []int, n int) {
	n %= len(s) // Handle rotations greater than slice length

	// Reverse the entire array
	reverse(s, 0, len(s)-1)

	// Reverse the first n elements
	reverse(s, 0, n-1)

	// Reverse the last len(s) - n elements
	reverse(s, n, len(s)-1)
}

func reverse(s []int, start, end int) {
	for i, j := start, end; i < j; i, j = i+1, j-1 {
		s[i], s[j] = s[j], s[i]
	}
}

File: web/src/features/matches/domain/generateRounds.ts
// Port of endpoint/schedule/internal/generate_rounds.go (pure logic, no I/O).
//
// Generates the round-robin matches within each group and the knockout bracket
// for a tournament, in place on the passed Tournament and returned. This
// replaces the legacy `POST /api/generateRounds` fetch with a synchronous local
// computation that produces the same `category.groups[].rounds` and
// `category.knockoutRounds` shape the Go endpoint returned.
//
// Field mapping (Go struct → TS type, from model/model.go vs @/shared/model):
//   Entry1Idx → entry1Idx, Entry2Idx → entry2Idx, DurationMinutes → durationMinutes,
//   Round → round, Matches → matches, EntriesIdx → entriesIdx,
//   KnockoutRounds → knockoutRounds, ShortName → shortName.
//   EntryByeIdx = -2, EntryEmptyIdx = -1 (same constants in @/shared/model).
//
// Go `panic`s on invalid generation become thrown `Error`s here; the UI is
// expected to catch them. Dead helpers from the Go file
// (getRoundPlayersIndicesWithRotation, generateSlice, rotateInPlace, reverse)
// are intentionally NOT ported — they are not on the active path.

import {
  EntryByeIdx,
  EntryEmptyIdx,
  type Match,
  type Group,
  type KnockoutRound,
  type Tournament
} from '@/shared/model'

// generateRoundsForTournament populates every category's group round-robins and
// knockout brackets on `tournament` and returns it. Mirrors GenerateRoundsForTournament.
export function generateRoundsForTournament(tournament: Tournament): Tournament {
  for (let i = 0; i < tournament.categories.length; i++) {
    const category = tournament.categories[i]

    for (let g = 0; g < category.groups.length; g++) {
      const rounds = generateGroupRounds(category.groups[g].entriesIdx, category.durationMinutes)
      if (category.groups[g].rounds.length === 0) {
        category.groups[g].rounds = rounds
      } else {
        if (category.groups[g].rounds.length !== rounds.length) {
          throw new Error(`number of rounds for group ${g + 1} is not equal`)
        }
        for (let j = 0; j < category.groups[g].rounds.length; j++) {
          for (let k = 0; k < category.groups[g].rounds[j].length; k++) {
            category.groups[g].rounds[j][k].entry1Idx = rounds[j][k].entry1Idx
            category.groups[g].rounds[j][k].entry2Idx = rounds[j][k].entry2Idx
          }
        }
      }
    }

    // Always generate knockout rounds based on the current NumQualifiedPerGroup.
    const koRounds = generateKnockoutRounds(category.groups, category.numQualifiedPerGroup)
    if (koRounds.length !== category.knockoutRounds.length) {
      category.knockoutRounds = koRounds
    } else {
      for (let j = 0; j < category.knockoutRounds.length; j++) {
        if (category.knockoutRounds[j].matches.length !== koRounds[j].matches.length) {
          category.knockoutRounds[j].matches = koRounds[j].matches
        } else {
          for (let k = 0; k < category.knockoutRounds[j].matches.length; k++) {
            category.knockoutRounds[j].matches[k].entry1Idx = koRounds[j].matches[k].entry1Idx
            category.knockoutRounds[j].matches[k].entry2Idx = koRounds[j].matches[k].entry2Idx
          }
        }
      }
    }
    tournament.categories[i] = category
  }
  return tournament
}

// swapRoundWithPlayersToEnd moves the round containing the (player1, player2)
// match to the final slot — port of the eponymous Go helper.
function swapRoundWithPlayersToEnd(rounds: Match[][], player1: number, player2: number): void {
  let roundIdx = -1
  for (let i = 0; i < rounds.length; i++) {
    if (i === rounds.length - 1) {
      continue
    }
    if (roundContains(rounds[i], player1, player2)) {
      roundIdx = i
    }
  }
  if (roundIdx >= 0) {
    const last = rounds.length - 1
    ;[rounds[roundIdx], rounds[last]] = [rounds[last], rounds[roundIdx]]
  }
}

function roundContains(round: Match[], player1Idx: number, player2Idx: number): boolean {
  for (const match of round) {
    if (match.entry1Idx === player1Idx && match.entry2Idx === player2Idx) {
      return true
    }
    if (match.entry1Idx === player2Idx && match.entry2Idx === player1Idx) {
      return true
    }
  }
  return false
}

function isRoundValid(rounds: Match[][], numMatches: number, numMatchesPerRound: number): boolean {
  let totalMatchCount = 0
  for (const matches of rounds) {
    if (matches.length !== numMatchesPerRound) {
      return false
    }
    totalMatchCount += matches.length
  }
  return totalMatchCount === numMatches
}

// nextPowerOfTwo returns the smallest power of two >= x.
// Port of the Go bit-twiddling implementation. (For the 32-bit JS operands used
// here the `>> 32` step is a no-op; this stays correct for all realistic player
// counts up to 2^31, which exceeds any tournament size.)
export function nextPowerOfTwo(x: number): number {
  if (x <= 1) {
    return 1
  }
  x--
  x |= x >> 1
  x |= x >> 2
  x |= x >> 4
  x |= x >> 8
  x |= x >> 16
  x |= x >> 32
  return x + 1
}

// generateKnockoutRounds builds the empty bracket (all EntryEmptyIdx matches)
// down to the final, sized to the next power of two over the qualified players.
export function generateKnockoutRounds(
  groups: Group[],
  numQualifiedPerGroup: number
): KnockoutRound[] {
  for (const group of groups) {
    if (group.entriesIdx.length < numQualifiedPerGroup) {
      throw new Error('not enough players')
    }
  }

  const qualifiedPlayersNum = groups.length * numQualifiedPerGroup
  const firstRound = nextPowerOfTwo(qualifiedPlayersNum)
  const numByes = firstRound - qualifiedPlayersNum
  const numMatches = firstRound / 2 - numByes

  const koRounds: KnockoutRound[] = []

  for (let round = firstRound; round >= 2; round = Math.floor(round / 2)) {
    const matchesCount = round === firstRound ? numMatches : round / 2
    const matches: Match[] = Array.from({ length: matchesCount }, () => ({
      entry1Idx: EntryEmptyIdx,
      entry2Idx: EntryEmptyIdx,
      datetime: '',
      durationMinutes: 0,
      table: ''
    }))
    koRounds.push({ round, matches })
  }

  return koRounds
}

// generateGroupRounds produces a complete round-robin for one group's entries
// using the circle method with the "bouncing" boundary reflection. Odd entry
// counts get a bye appended (EntryByeIdx). Mirrors generateGroupRounds.
export function generateGroupRounds(entriesIdx: number[], matchDurationMinutes: number): Match[][] {
  if (entriesIdx.length < 2) {
    return []
  }

  // Copy so we never mutate the caller's entriesIdx slice (Go's append returns a
  // new slice header; slice + push here gives the same isolation).
  const players = entriesIdx.slice()
  let numPlayers = players.length
  const numMatches = (numPlayers * (numPlayers - 1)) / 2
  // Integer division (Go used integer `/`; the JS port must floor). For even
  // group sizes this is identical to plain division, so even-size golden
  // outputs are unchanged. For ODD sizes a bye is appended below, and flooring
  // keeps numMatchesPerRound/numRounds whole (e.g. 3 players -> 1 match/round
  // over 3 rounds) so isRoundValid() does not reject the valid schedule.
  const numMatchesPerRound = Math.floor(numPlayers / 2)
  const numRounds = Math.floor(numMatches / numMatchesPerRound)

  if (numPlayers % 2 === 1) {
    players.push(EntryByeIdx)
    numPlayers++
  }

  const rounds: Match[][] = Array.from({ length: numRounds }, (_, r) =>
    getRoundMatches(r, players, matchDurationMinutes)
  )

  if (!isRoundValid(rounds, numMatches, numMatchesPerRound)) {
    throw new Error('generateGroupRounds encounter error')
  }

  if (players.length > 2) {
    swapRoundWithPlayersToEnd(rounds, players[1], players[2])
  }
  return rounds
}

// getRoundMatches builds the list of matches for a single round from the player
// indices produced by getRoundPlayersIndices. A bye entry is skipped (no match).
// Port of getRoundMatches; the in/out `indices` buffer is allocated locally.
export function getRoundMatches(
  round: number,
  entriesIdx: number[],
  matchDurationMinutes: number
): Match[] {
  const indices = getRoundPlayersIndices(round, entriesIdx.length)
  const matches: Match[] = []

  for (let i = 0; i < indices.length; i += 2) {
    let ind1 = indices[i]
    let ind2 = indices[i + 1]
    if (ind2 < ind1) {
      ;[ind1, ind2] = [ind2, ind1]
    }
    const p1 = entriesIdx[ind1]
    const p2 = entriesIdx[ind2]
    if (p1 === EntryByeIdx || p2 === EntryByeIdx) {
      continue
    }
    matches.push({
      entry1Idx: p1,
      entry2Idx: p2,
      datetime: '',
      durationMinutes: matchDurationMinutes,
      table: ''
    })
  }
  return matches
}

// getRoundPlayersIndices computes the player indices for a given round in a
// round-robin draw via the "bouncing" boundary reflection. res[0] is always 0
// (player 0 stays fixed); the rest rotate with reflection at both boundaries.
// Port of getRoundPlayersIndices verbatim — this is the subtle part.
export function getRoundPlayersIndices(round: number, numPlayers: number): number[] {
  if (numPlayers % 2 === 1) {
    throw new Error('num of players should be even')
  }
  if (round + 1 >= numPlayers) {
    throw new Error('invalid number of rounds or numPlayers')
  }

  const res: number[] = new Array(numPlayers)
  res[0] = 0

  // Iterate through all players except player 0 (who stays fixed at position 0)
  for (let i = 1; i < numPlayers; i++) {
    let newPos: number

    if (i % 2 === 0) {
      // For players at even positions (2, 4, 6, etc.) — clockwise rotation
      newPos = i + 2 * round
      if (newPos >= numPlayers) {
        // Bounce off the upper boundary
        newPos = 2 * numPlayers - newPos - 1
        if (newPos < 0) {
          newPos = -(newPos - 1)
        }
      }
    } else {
      // For players at odd positions (1, 3, 5, etc.) — counterclockwise rotation
      newPos = i - 2 * round
      if (newPos < 0) {
        // Bounce off the lower boundary
        newPos = -(newPos - 1)
        if (newPos >= numPlayers) {
          newPos = 2 * numPlayers - newPos - 1
        }
      }
    }

    res[i] = newPos
  }
  return res
}
